adafruit_azureiot.py

```diff
@@ -0,0 +1,245 @@
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+"""
+`adafruit_azureiot`
+================================================================================
+
+Microsoft Azure IoT for CircuitPython
+
+* Author(s): Brent Rubell
+
+Implementation Notes
+--------------------
+
+**Software and Dependencies:**
+
+* Adafruit CircuitPython firmware for the supported boards:
+  https://github.com/adafruit/circuitpython/releases
+
+* Adafruit's Bus Device library: https://github.com/adafruit/Adafruit_CircuitPython_BusDevice
+* Adafruit's ESP32SPI library: https://github.com/adafruit/Adafruit_CircuitPython_ESP32SPI
+"""
+
+__version__ = "0.0.0-auto.0"
+__repo__ = "https://github.com/adafruit/Adafruit_CircuitPython_AzureIoT.git"
+
+AZ_API_VER = "2018-06-30" # Azure URI API Version Identifier
+AZURE_HTTP_ERROR_CODES = [400, 401, 404, 403, 412, 429, 500] # Azure HTTP Status Codes
+
+class IOT_Hub:
+    """
+    Provides access to a Microsoft Azure IoT Hub.
+    https://docs.microsoft.com/en-us/rest/api/iothub/
+    """
+    def __init__(self, wifi_manager, iot_hub_name, sas_token, device_id):
+        """ Creates an instance of an Azure IoT Hub Client.
+        :param wifi_manager: WiFiManager object from ESPSPI_WiFiManager.
+        :param str iot_hub_name: Name of your IoT Hub.
+        :param str sas_token: Azure IoT Hub SAS Token Identifier.
+        :param str device_id: Unique Azure IoT Device Identifier.
+        """
+        _wifi_type = str(type(wifi_manager))
+        if 'ESPSPI_WiFiManager' in _wifi_type:
+            self._wifi = wifi_manager
+        else:
+            raise TypeError("This library requires a WiFiManager object.")
+        self._iot_hub_url = "https://{0}.azure-devices.net".format(iot_hub_name)
+        self._sas_token = sas_token
+        self._device_id = device_id
+        self._azure_header = {"Authorization":self._sas_token}
+
+    @property
+    def device_id(self):
+        """Returns the current device identifier"""
+        return self._device_id
+
+    @device_id.setter
+    def device_id(self, device_identifier):
+        """Sets the current device identifier
+        :param str device_identifier: Unique device identifier.
+        """
+        self._device_id = device_identifier
+
+    @staticmethod
+    def _parse_http_status(status_code, status_reason):
+        """Parses status code, throws error based on Azure IoT Common Error Codes.
+        :param int status_code: HTTP status code.
+        :param str status_reason: Description of HTTP status.
+        """
+        for error in AZURE_HTTP_ERROR_CODES:
+            if error == status_code:
+                raise TypeError("Error {0}: {1}".format(status_code, status_reason))
+
+    # Cloud-to-Device Messaging
+    def get_hub_message(self):
+        """Returns a message from a Microsoft Azure IoT Hub (Cloud-to-Device), or -1
+        if the message queue is empty.
```

Comment:
One tiny thing here would be to return None when empty rather than -1.

Reply:
Good idea, added this in PR https://github.com/adafruit/Adafruit_CircuitPython_AzureIoT/pull/3